plugins/outputs/mqtt/mqtt.go

```diff
@@ -142,58 +145,67 @@ func (m *MQTT) Write(metrics []telegraf.Metric) error {
 }
 
 func (m *MQTT) publish(topic string, body []byte) error {
-	token := m.client.Publish(topic, byte(m.QoS), m.Retain, body)
-	token.WaitTimeout(time.Duration(m.Timeout))
-	if token.Error() != nil {
-		return token.Error()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(m.Timeout))
+	defer cancel()
+	_, err := m.client.Publish(ctx, &paho.Publish{
+		Topic:   topic,
+		QoS:     byte(m.QoS),
+		Retain:  m.Retain,
+		Payload: body,
+	})
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
-func (m *MQTT) createOpts() (*paho.ClientOptions, error) {
-	opts := paho.NewClientOptions()
-	opts.KeepAlive = m.KeepAlive
-
-	if m.Timeout < config.Duration(time.Second) {
-		m.Timeout = config.Duration(5 * time.Second)
-	}
-	opts.WriteTimeout = time.Duration(m.Timeout)
+func (m *MQTT) createOpts() (*autopaho.ClientConfig, error) {
+	opts := &autopaho.ClientConfig{}
 
 	if m.ClientID != "" {
-		opts.SetClientID(m.ClientID)
+		opts.ClientID = m.ClientID
 	} else {
-		opts.SetClientID("Telegraf-Output-" + internal.RandomString(5))
+		opts.ClientID = "Telegraf-Output-" + internal.RandomString(5)
 	}
 
+	user := m.Username
+	pass := m.Password
+	opts.SetUsernamePassword(user, []byte(pass))
+
 	tlsCfg, err := m.ClientConfig.TLSConfig()
 	if err != nil {
 		return nil, err
 	}
 
-	scheme := "tcp"
 	if tlsCfg != nil {
-		scheme = "ssl"
-		opts.SetTLSConfig(tlsCfg)
-	}
-
-	user := m.Username
-	if user != "" {
-		opts.SetUsername(user)
-	}
-	password := m.Password
-	if password != "" {
-		opts.SetPassword(password)
+		opts.TlsCfg = tlsCfg
 	}
 
 	if len(m.Servers) == 0 {
 		return opts, fmt.Errorf("could not get host informations")
 	}
+
+	brokers := make([]*url.URL, 0)
 	for _, host := range m.Servers {
-		server := fmt.Sprintf("%s://%s", scheme, host)
+		u, err := url.Parse(host)
+		if u.Scheme == "" {
+			if opts.TlsCfg != nil {
+				u.Scheme = "mqtts"
+			} else {
+				u.Scheme = "mqtt"
+			}
+		}
+		if err != nil {
+			return opts, fmt.Errorf("could not create url: %s", err)
+		}
+		brokers = append(brokers, u)
+	}
+	opts.BrokerUrls = brokers
 
-		opts.AddBroker(server)
+	opts.KeepAlive = uint16(m.KeepAlive)
+	if m.Timeout < config.Duration(time.Second) {
+		m.Timeout = config.Duration(5 * time.Second)
```

Comment:
Shouldn't we allow for smaller timeouts? What about setting the default in `init()` (lowercase) as this can then still be overwritten by the user?

Reply:
I was reusing the original logic here https://github.com/influxdata/telegraf/blob/master/plugins/outputs/mqtt/mqtt.go#L157-L159

I am fine with just setting `m.Timeout` directly w/o any logic. Thoughts?

Reply:
Keep it for now, but I see this pattern a lot and it should be fixed.